🏗 Add experiment test coverage to Travis

estherkim opened the pull request:
For push builds, build a dist output for every experiment in `experiments-config.json`. Run integration tests for these as part of the e2e job, now renamed `End to End, Experiment Tests`

estherkim commented:
@zhouyx @jonathantyng-amp - just saw https://github.com/ampproject/amphtml/pull/23604, will wait til it goes through and adjust this PR as needed.

zhouyx commented on build-system/global-configs/experiments-config.json:
```diff
@@ -1,7 +1,7 @@
 {
   "experimentA": {
     "name": "ANALYTICS_VENDOR_SPLIT",
-    "command": "gulp dist --defineExperimentConstant=ANALYTICS_VENDOR_SPLIT",
```
hmmmm we need this `command` to build experiment bundle in release. Can we add --fortesting automatically, or we can introduce another test command? WDYT?

estherkim replied:
ooh I see, undid the change and did this instead
https://github.com/ampproject/amphtml/blob/4de59615a9c50d8f6da1ed0d0878d6771d56fa7a/build-system/pr-check/dist-experiment.js#L42-L45

zhouyx replied:
Great! That sounds good!

zhouyx commented:
FYI #23604 merged. Thanks

rsimha commented on build-system/pr-check/dist-bundle-size.js:
```diff
@@ -55,6 +69,7 @@ async function main() {
     timedExecOrDie('gulp dist --fortesting');
     timedExecOrDie('gulp bundle-size --on_push_build');
     uploadDistOutput(FILENAME);
+    buildAndUploadExperiments_();
```
I'm concerned this will double the running time of the `build` stage of push builds, and potentially slow down the release process. Would it be possible to do this in a separate parallel job?

estherkim replied:
Yeah good idea, here's what it could look like on push builds. 
![image](https://user-images.githubusercontent.com/44627152/62317070-dbdf0200-b466-11e9-9224-5c89d3350d18.png)

estherkim replied:
The job duration is just for 1 experiment, though. I'll have to figure out how to run multiple in parallel.

rsimha replied:
Nice! New matrix LGTM.

Yeah, fitting in more experiments is an interesting problem. Possible ideas:

- Build and test experiments only for canary and release branches, and for PRs that modify experimental code
- Optimize experimental builds to incrementally rebuild just the extensions that contain experimental code, and reuse the build output (not sure how feasible this is)

estherkim replied:
So I figured out run test suites in parallel on 1 vm using multiple node child processes! It works fine with e2e tests after removing the hardcoded 8000 port in some places. Unfortunately it's a lot hairier for integration tests so I'm not running those yet.

For running builds in parallel, still haven't figured that out / whether it's feasible.

rsimha commented on build-system/pr-check/e2e-experiment-tests.js:
```diff
@@ -22,28 +22,47 @@
     downloadDistOutput(FILENAME);
     timedExecOrDie('gulp update-packages');
     timedExecOrDie('gulp e2e --nobuild --headless');
+    await runExperimentTests_();
```
Curious: What is the new running time of the E2E job? (I'd imagine it's still shorter than or close to the local / remote tests jobs?)

estherkim commented:
> Woo! Thank you @estherkim
> 
> I have a question about e2e tests. Does it cover both the e2e test and the integration test?

It doesn't right now but it could. I can add it and see how long it takes.

zhouyx commented:
> I have a question about e2e tests. Does it cover both the e2e test and the integration test?
It doesn't right now but it could. I can add it and see how long it takes.

I thought we only added e2e test coverage, that's why i asked the question. But looked at the code again, looks like we are running integration test plus e2e test for experiment build 😄. Please correct me if that's not true.

estherkim commented:
I changed it to run both! Before it was just integration tests but it was using the same Travis job as the e2e tests for normal dist.

jonathantyng-amp commented:
Will need this PR (https://github.com/ampproject/amphtml/pull/23949) for the `ANALYTICS_VENDOR_SPLIT` experiment integration tests to pass. A test JSON that the integration tests relied on was missing.

estherkim commented:
Ok cool - discussed with @zhouyx that running tests in their own VMs are the best, esp now that there will only be 3 experiments max. Removed parallel logic and added a third "Experiment" stage on Travis - https://travis-ci.org/ampproject/amphtml/builds/574443414?utm_source=github_status&utm_medium=notification

@zhouyx - if an experiment test fails on master, should the entire build fail?

zhouyx commented:
The three new experiments stage looks great! 

>if an experiment test fails on master, should the entire build fail?

I do think so, otherwise we would never notice the test failure. However I don't know how flaky our integration tests are. What are the other options? Thanks

estherkim commented:
What if we ran blocking experiment tests on PR builds, but only if a PR had code changes related to any of the experiments? Is it easy to determine this?

zhouyx commented:
It would be difficult to detect. Especially we are thinking about moving the singlepass experiment to one of the three experiments. 

I think running the test only for master build should be sufficient.

estherkim commented:
> I think running the test only for master build should be sufficient.

Sounds good - made that change. This is ready for another review.
